kernel-rs/src/lock/mod.rs

```diff
@@ -0,0 +1,156 @@
+mod spinlock;
+mod spinlock_protected;
+
+pub use sleepablelock::{Sleepablelock, SleepablelockGuard};
+pub use sleeplock::{Sleeplock, SleeplockGuard};
+pub use spinlock::{pop_off, push_off, Spinlock, SpinlockGuard};
+pub use spinlock_protected::{SpinlockProtected, SpinlockProtectedGuard};
+
+pub trait RawLock {
+    /// Acquires the lock.
+    fn acquire(&self);
+    /// Releases the lock.
+    fn release(&self);
+    /// Check whether this cpu is holding the lock.
+    fn holding(&self) -> bool;
+}
+
+/// Locks that provide mutual exclusion and has its own `RawLock`.
+pub struct Lock<R: RawLock, T> {
+    lock: R,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl<R: RawLock, T: Send> Sync for Lock<R, T> {}
+
+/// Guards that guarantee exclusive mutable access to the lock's inner data.
+pub struct Guard<'s, R: RawLock, T> {
+    lock: &'s Lock<R, T>,
+    _marker: PhantomData<*const ()>,
+}
+
+// Do not implement Send; lock must be unlocked by the CPU that acquired it.
+unsafe impl<'s, R: RawLock, T: Sync> Sync for Guard<'s, R, T> {}
+
+impl<R: RawLock, T> Lock<R, T> {
+    /// Acquires the lock and returns the lock guard.
+    pub fn lock(&self) -> Guard<'_, R, T> {
+        self.lock.acquire();
+
+        Guard {
+            lock: self,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns a mutable reference to the inner data.
+    /// The returned pointer is valid until this lock is moved or dropped.
+    /// The caller must ensure that accessing the pointer does not incur race.
+    /// Also, if `T: !Unpin`, the caller must not move the data using the pointer.
+    pub fn get_mut_raw(&self) -> *mut T {
+        self.data.get()
+    }
+
+    /// Returns a pinned mutable reference to the inner data.
+    pub fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut T> {
+        // Safe since for `T: !Unpin`, we only provide pinned references and don't move `T`.
+        unsafe { Pin::new_unchecked(&mut *self.get_mut_raw()) }
+    }
+
+    /// Returns a mutable reference to the inner data.
+    pub fn get_mut(&mut self) -> &mut T
+    where
+        T: Unpin,
+    {
+        // Safe since we have a mutable reference of the lock.
+        unsafe { &mut *self.get_mut_raw() }
+    }
+
+    /// Consumes the lock and returns the inner data.
+    pub fn into_inner(self) -> T
+    where
+        T: Unpin,
+    {
+        self.data.into_inner()
+    }
+
+    /// Unlock the lock.
+    ///
+    /// # Safety
+    ///
+    /// Use this only when we acquired the lock but did `mem::forget()` to the guard.
+    pub unsafe fn unlock(&self) {
+        self.lock.release();
+    }
+}
+
+impl<R: RawLock, T> Guard<'_, R, T> {
+    pub fn reacquire_after<F, U>(&mut self, f: F) -> U
+    where
+        F: FnOnce() -> U,
+    {
+        self.lock.lock.release();
+        let result = f();
+        self.lock.lock.acquire();
+        result
+    }
+
+    /// Returns a pinned mutable reference to the inner data.
+    pub fn get_pin_mut(&mut self) -> Pin<&mut T> {
```

Comment:
저도 조금 생각해봤는데,
* `Lock::get_pin_mut()`를 사용한 경우 lifetime이 `&'s Lock` -> `Pin<&'s mut T>`,
* `Lock::lock()`을 사용한 경우 lifetime이 `&'s Lock` -> `Guard<'s, R, T>`, 그리고 이후 `Guard::get_pin_mut()`을 사용하면 `&'t mut Guard<'s, R, T>` -> `Pin<&'t mut T>`.

따라서, 어느 경우에든 `Pin<&mut T>`의 lifetime동안에는 `Lock`이 immutable borrowed되었으므로, move가 static analysis에 의하여 금지되지 않을까 싶습니다.

Reply:
실제로 다음과 같은 code는 static analysis 단계에서 금지되는 것 같습니다.
```Rust
fn test() {
    let mut lock = Spinlock::new("test", 100);
    let mut guard = lock.lock();
    let pin = guard.get_pin_mut();

    let mut lock2 = Spinlock::new("test2", 200);
    core::mem::swap(&mut lock, &mut lock2);
    assert!(*pin == 100);
}
```

Reply:
제가 걱정한 부분은 "나중에 guard가 drop된 후에 move가 일어나는 경우"입니다.

```rust
fn test() {
    let mut lock = Spinlock::new("test", 100);
    let mut guard = lock.lock();
    let pin = guard.get_pin_mut();
    // use pin
    drop(guard);

    let mut lock2 = Spinlock::new("test2", 200);
    core::mem::swap(&mut lock, &mut lock2);
}
```

`Pin::new_unchecked`의 safety condition 설명에 다음과 같은 문장이 있는데, 위와 같은 코드는 이 조건을 깨는 것이 아닌가 해서요.

> This constructor is unsafe because we cannot guarantee that the data pointed to by pointer is pinned, meaning that the data will not be moved or its storage invalidated until it gets dropped.

Reply:
네, 사실 이렇게 되면 "once pinned, must remain pinned forever"라는 contract을 깨는게 맞습니다.

다만, 제 생각에는 `Lock::lock()`의 signature를 `fn lock(&self)`에서 `fn lock(self: Pin<&Self>)`로 바꾸면 `Lock::new_unchecked()`를 굳이 추가하지 않고도 해결할 수 있을 것 같습니다. (참고로, 이러면 `Pin contract`에 대한 책임을 caller에게 돌리는 셈이 됩니다.) 그러나, 이러면 readability가 안좋아질 수 있으므로, 차라리 `Lock::new_unchecked()`를 다시 추가하는게 나을 수 있습니다.

Reply:
저는 `T: !Unpin`일 때 lock을 acquire하려면 lock이 pin되어 있어야만 하는 것(즉, `fn lock(self: Pin<&Self>)`)이 자연스러운 설계인 것 같습니다.

Reply:
`fn lock()`의 signature를 위와 같이 변경해봤는데, 변경해야하는 부분이 매우 많이 생기는 것 같습니다. (commit은 안했습니다.) 물론 대부분의 경우, 단순히 `&self`를 `self: Pin<&Self>`로 바꾸기만 하면 되긴 하지만, 문제는 그 수가 매우 많습니다. 참고로, `MruArena`는 이미 `!Unpin`이고, `ArrayArena`는 PR #434 에서 `!Unpin`으로 바뀔 예정이기에, `Arena` trait와 관련된 function들의 signature를 모두 바꿔야합니다.

이 부분에 대해서는 내일 sync up에서 조금 이야기해보면 좋을 것 같습니다.

Reply:
이 부분은 다른 PR에서 해결하는 게 좋을 것 같습니다. Issue #439 를 추가했습니다.